all-of-us/raw-data-repository
# [DA-2140] Consent report tool:  Add CSV file generation (for PTSC)

## Resolves *[DA-2140](https://precisionmedicineinitiative.atlassian.net/browse/DA-2140)*


## Description of changes/additions
This adds the CSV error file generation that will be sent to PTSC when consent validations fail to the `consent-report` local tool.  For now, the file is created in the local directory where the tool is run.  Expectation is we will upload the files manually for now into service desk tickets for PTSC.   

Other changes:

- Minor refactoring of the SQL to limit the regular daily report to only include validations on new consents uploaded or flagged as missing on the designated report date (vs. pulling in potentially large numbers of retrospective validations that happened to be performed on the same day)
- fixed bugs in the error filtering logic, partially related to presence of NaN/null values in dataframe (now filled with zeros)
- Add additional report options that can capture all validation errors from a given date (including retrospective validations) or that can report on all unresolved (NEEDS_CORRECTING) consent errors, regardless of date
- Command line options to selectively suppress generating either the CSV file or the gsheets file, if only one is needed
- Refactor how the error counts worksheet row data is populated, to cut down on gsheets API calls (which are rate-limited)
- Add some report notes included in the worksheet to clarify for stakeholders limitations on the validation counts

## Tests
Manual testing of reports.  To test operation of this code in your local environment:
- Pull down branch and update your virtual environment packages (`pip-install -r requirements.txt`)
- Create a private google sheets doc and give access to the prod SA account email
-  Use the Google doc id/key (from the share link) with the `--doc-id` argument to run the tool:
`python -m tools consent-report --project <prod project> --service-account <prod SA> --doc-id <your doc key>`

Note that to re-run the report with the same report date, you have to delete the previously created tab for that date from the worksheet first

Use `python -m tools consent-report --help` to explore other options.
